waspc/src/Wasp/Generator/Setup.hs

```diff
@@ -1,22 +1,82 @@
 import Wasp.Generator.Job.IO (printPrefixedJobMessage)
-import Wasp.Generator.ServerGenerator.Setup (setupServer)
-import Wasp.Generator.WebAppGenerator.Setup (setupWebApp)
+import Wasp.Generator.Monad (GeneratorError (..), GeneratorWarning (..))
+import qualified Wasp.Generator.PackageJsonGenerator as PJG
+import Wasp.Generator.ServerGenerator as SG
+import qualified Wasp.Generator.ServerGenerator.Setup as SSetup
+import Wasp.Generator.WebAppGenerator as WG
+import qualified Wasp.Generator.WebAppGenerator.Setup as WSetup
 
-setup :: Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
-setup projectDir = do
+runSetup :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO ([GeneratorWarning], [GeneratorError])
+runSetup spec dstDir = do
+  (npmInstallWarnings, npmInstallErrors) <- runNpmInstall spec dstDir
+  (dbGeneratorWarnings, dbGeneratorErrors) <- DbGenerator.postWriteDbGeneratorActions spec dstDir
+  return (npmInstallWarnings ++ dbGeneratorWarnings, npmInstallErrors ++ dbGeneratorErrors)
+
+-- | Run npm install if needed
+runNpmInstall :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO ([GeneratorWarning], [GeneratorError])
+runNpmInstall spec dstDir = do
+  let appSpecAllPackageJsonDependencies = PJG.getAllPackageJsonDependencies spec SG.waspPackageJsonDependencies WG.waspPackageJsonDependencies
+  case appSpecAllPackageJsonDependencies of
+    Nothing -> return ([], [GenericGeneratorError "npm install failed due to conflict between user and wasp dependencies."])
+    Just allDependencies -> executeNpmInstall allDependencies dstDir
+
+-- Run npm install for desired AppSpec dependencies if needed
+executeNpmInstall :: PJG.AllPackageJsonDependencies -> Path' Abs (Dir ProjectRootDir) -> IO ([GeneratorWarning], [GeneratorError])
+executeNpmInstall appSpecAllPackageJsonDependencies dstDir = do
+  needed <- isNpmInstallNeeded appSpecAllPackageJsonDependencies dstDir
+  npmInstallResult <-
+    if needed
+      then installNpmDependencies dstDir
+      else return $ Right ()
+  case npmInstallResult of
+    Left npmInstallError -> do
+      return ([], [GenericGeneratorError $ "npm install failed: " ++ npmInstallError])
+    Right () -> do
+      -- on successful npm install, record what we wanted to install
+      B.writeFile dependenciesInstalledFp (Aeson.encode appSpecAllPackageJsonDependencies)
+      return ([], [])
+  where
+    dependenciesInstalledFp = SP.fromAbsFile $ dstDir </> npmDependenciesFileInProjectRootDir
+
+isNpmInstallNeeded :: PJG.AllPackageJsonDependencies -> Path' Abs (Dir ProjectRootDir) -> IO Bool
+isNpmInstallNeeded appSpecAllPackageJsonDependencies dstDir = do
+  installedAllPackageJsonDependencies <- loadInstalledAllPackageJsonDependencies dstDir
+  return $ Just appSpecAllPackageJsonDependencies /= installedAllPackageJsonDependencies
+
+npmDependenciesFileInProjectRootDir :: Path' (Rel ProjectRootDir) File'
+npmDependenciesFileInProjectRootDir = [relfile|npmDependencies.json|]
```

Comment:
I've done various renamings and added a comment about the meta directory. PackageJson -> Npm throughout which is shorter and just as clear.

Reply:
I am not sure how keen you will be to discuss naming after you did all those changes, but I still have some of the similar thoughts I expressed earlier above.

Specifically, name for the record that contains both dependencies and dev dependencies.

Right now you name it `NpmDependencies`.
But, when I see `NpmDependencies`, I imagine that is a list of dependencies, not a union of non-dev and dev npm dependencies.

npm dependency -> dependency on npm

npm dependencies -> multiple npm dependencies

So what is this mix of dev and non-dev npm dependencies? Why does that even exist? I believe the answer to that is: "npm package". So npm package is a distributable unit that defines npm dependencies it depends on. And it separates those dependencies into dev and non-dev deps. If I am correct, this concept of dev and non-dev npm deps doesn't exist outside of context of npm package.

So, if we have a record of dev and non-dev npm deps, the best name for it would be `NpmPackageDependencies`.

And FullStackNpmDependencies -> we could name it FullStackNpmPackageDependencies.

This we avoid this weird situation of NpmDependencies containing dependencies, but it already is dependencies -> weird, now what one would expect.

Reply:
Regarding all this naming stuff -> just to be clear, I am ok with you leaving it as it is, if you feel like that is better. Just wanted to offer my thoughts on the matter since we are already looking into it.

Reply:
I used to call it `PackageJsonDependencies`. So now I feel I could just name it back again.

Reply:
Well "Json" is certainly redundant I think. That is just file format.

"Package" has a meaning.

So I don't think naming it back would be a step forward.

Reply:
But package could be any package. :)